Doc: Introduce snmp-integration for Technical Preview (#1648)
elastic/logstash-docs: docs/plugins/integrations/snmp.asciidoc

END - GENERATED VARIABLES, DO NOT EDIT!
///////////////////////////////////////////

// include::{include_path}/plugin_header.asciidoc[]

experimental[]

.Technical Preview
****

> The new `integration-snmp` plugin, and its component plugins--`input-snmp` and `input-snmptrap`--are available in _Technical Preview_ at {ls} 8.14.0.

Author: Note to self @karenzone: Get exact versioning right, and determine how we want to handle this for minor and patch releases (lockfile).

****
The new `integration-snmp` plugin, and its component plugins--`input-snmp` and `input-snmptrap`--are available in _Technical Preview_ and can be installed on the latest Logstash 7.x and 8.x versions.

Current 1.x versions of the `input-snmp` plugin are bundled with {ls} by default, and will soon be replaced by the snmp input plugin contained in this integration.
(If you do want to opt into the Technical Preview for the `integration-snmp` plugin, run `bin/logstash-plugin install logstash-integration-snmp`.)

Be aware of behavioral and mapping differences between current stand-alone plugins and the new versions included in the `integration-snmp`. 
The information in this topic can help. 
****

==== Description

The SNMP integration plugin includes:

* SNMP input plugin
* SNMPtrap input plugin

// Uncomment and replace with above bullets after generated the initial versions of the target files
// - {logstash-ref}/plugins-input-snmp.html[SNMP input plugin]
// - {logstash-ref}/plugins-input-snmptrap.html[Snmptrap input plugin]

The new `logstash-integration-snmp` plugin combines the 
`logstash-input-snmp` and `logstash-input-snmptrap` plugins into a single Ruby gem.

> This integrated plugin package provides better alignment in behaviour when it comes to SNMP processing, better resource management, easier package maintenance, and a smaller installation footprint. 
> The individual plugins now share the same code base and have been refactored to leverage the latest version of https://www.snmp4j.org/[SNMP4j].

Author: ```suggestion
This integrated plugin package provides better alignment in SNMP processing behavior, better resource management, easier package maintenance, and a smaller installation footprint. 
The individual plugins now share the same code base and have been refactored to leverage the latest version of https://www.snmp4j.org/[SNMP4j].
```

[id="plugins-{type}s-{plugin}-migration"]
==== Migrating to `logstash-integration-snmp` from individual plugins

You'll retain the functionality of existing stand-alone plugins, but with a more consistent experience and expanded features. 
In this section, we'll note mapping and behavioral changes, and explain how to preserve current behavior if needed.

// ToDo: Add guidance and set expectations for when snmp integration is installed by default

[id="plugins-{type}s-{plugin}-migration-input-snmp"]
===== Migration notes: `logstash-input-snmp` 

As a component of the new `logstash-integration-snmp` plugin, the `logstash-input-snmp` plugin offers the same 
capabilities as the stand-alone https://github.com/logstash-plugins/logstash-input-snmp[logstash-input-snmp]. 

> Given the underlying code refactor to better align snmp and snmptrap plugin behaviour you might need to address some changes depending on the use-case and how the ingested data is being handled through the pipeline.

Author: ```suggestion
Given the underlying code refactor to better align snmp and snmptrap plugin behavior, you might need to address some changes depending on your use case and how the ingested data is being handled through the pipeline.
```

=== SNMP Integration Plugin

* Plugin version: v4.0.0
* Released on: 2024-05-03
* https://github.com/logstash-plugins/logstash-integration-snmp/blob/main/CHANGELOG.md[Changelog]

experimental[]

.Technical Preview
****

> The new `integration-snmp` plugin, and its component plugins--`input-snmp` and `input-snmptrap`--are available in _Technical Preview_ at {ls} 8.14.0.

Reviewer: ```suggestion
The new `integration-snmp` plugin, and its component plugins--`input-snmp` and `input-snmptrap`--are available in _Technical Preview_ and can be installed on the latest Logstash 7.x and 8.x versions.
```

> Pre-4.0.0 versions of the `input-snmp` plugin are bundled with {ls} by default, and will not be upgraded to the snmp integration without your deliberate action. 

Reviewer: ```suggestion
Current 1.x versions of the `input-snmp` plugin are bundled with {ls} by default, and will soon be replaced by the snmp input plugin contained in this integration.
```

(If you do want to opt into the Technical Preview for the `integration-snmp` plugin, run `bin/logstash-plugin install logstash-integration-snmp`.)

Be aware of behavioral and mapping differences between current stand-alone plugins and the new versions included in the `integration-snmp`. 
The information in this topic can help. 
****

==== Description

The SNMP integration plugin includes:

* SNMP input plugin
* SNMPtrap input plugin

// Uncomment and replace with above bullets after generated the initial versions of the target files
// - {logstash-ref}/plugins-input-snmp.html[SNMP input plugin]
// - {logstash-ref}/plugins-input-snmptrap.html[Snmptrap input plugin]

> The new the `logstash-integration-snmp` plugin combines the 
> `logstash-input-snmp` and `logstash-input-snmptrap` plugins into one integrated plugin that encompasses
> the capabilities of both. 
> This integrated plugin package provides better alignment in snmp processing, better resource managenment, 
> easier package maintenance, and a smaller installation footprint. 

Reviewer: ```suggestion
The new `logstash-integration-snmp` plugin combines the 
`logstash-input-snmp` and `logstash-input-snmptrap` plugins into a single Ruby gem.
This integrated plugin package provides better alignment in behaviour when it comes to SNMP processing, better resource management, easier package maintenance, and a smaller installation footprint. 
The individual plugins now share the same code base and have been refactored to leverage the latest version of https://www.snmp4j.org/[SNMP4j].
```

[id="plugins-{type}s-{plugin}-migration"]
==== Migrating to `logstash-integration-snmp` from individual plugins

> You'll retain and expand the functionality of existing stand-alone plugins, but in a more compact, integrated package. 

Reviewer: ```suggestion
You'll retain the functionality of existing stand-alone plugins, but with a more consistent experience and expanded features. 
```

In this section, we'll note mapping and behavioral changes, and explain how to preserve current behavior if needed.

> [IMPORTANT] 
> --
> Uninstall the `logstash-input-snmp` and `logstash-input-snmptrap` plugins _before_ you install the `logstash-integration-snmp` plugin. 
> 
> * `bin/logstash-plugin remove logstash-input-snmp` +
> * `bin/logstash-plugin remove logstash-input-snmptrap`
> --

Reviewer: ```suggestion
```
This shouldn't be necessary as installing the snmp integration plugin uninstalls the individual plugins it provides.

// ToDo: Add guidance and set expectations for when snmp integration is installed by default

[id="plugins-{type}s-{plugin}-migration-input-snmp"]
===== Migration notes: `logstash-input-snmp` 

As a component of the new `logstash-integration-snmp` plugin, the `logstash-input-snmp` plugin offers the same 
capabilities as the stand-alone https://github.com/logstash-plugins/logstash-input-snmp[logstash-input-snmp]. 

> You might need to address some behavior changes depending on the use-case and how the ingested data is being handled through the pipeline.

Reviewer: ```suggestion
Given the underlying code refactor to better align snmp and snmptrap plugin behaviour you might need to address some changes depending on the use-case and how the ingested data is being handled through the pipeline.
```
